File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/tasklistview/TaskListViewLogic.java
package com.wiseassblog.samsaradayplanner.ui.tasklistview;

import com.wiseassblog.samsaradayplanner.common.BaseViewLogic;
import com.wiseassblog.samsaradayplanner.common.Continuation;
import com.wiseassblog.samsaradayplanner.domain.ITaskStorage;
import com.wiseassblog.samsaradayplanner.domain.Tasks;

public class TaskListViewLogic extends BaseViewLogic<TasksListViewEvent> {
    private final ITaskListViewContract.View view;
    private final ITaskListViewContract.ViewModel vm;
    private final ITaskStorage storage;

    public TaskListViewLogic(ITaskListViewContract.View view, ITaskListViewContract.ViewModel vm, ITaskStorage storage) {
        this.view = view;
        this.vm = vm;
        this.storage = storage;
    }


    @Override
    public void onViewEvent(TasksListViewEvent event) {
        switch (event.getEvent()) {
            case ON_START:
                onStart();
                break;
            case ON_LIST_ITEM_SELECTED:
                onItemSelected((int)event.getValue());
                break;
            case ON_BACK_PRESSED:
                onBackPressed();
                break;
        }
    }

    private void onBackPressed() {
        view.navigateToDayView();
    }

    private void onItemSelected(int position) {
        int id = vm.getTasks().get()[position].getTaskId();

        view.navigateToTaskView(id);
    }

    private void onDone() {

    }

    private void onStart() {
        storage.getTasks(
                new Continuation<Tasks>() {
                    @Override
                    public void onSuccess(Tasks result) {
                        vm.setTasks(result);
                        view.setTasks(result);
                    }

                    @Override
                    public void onException(Exception e) {
                        view.showMessage(e.getMessage());
                        view.navigateToDayView();
                    }
                }
        );
    }
}

File: app/src/main/java/com/wiseassblog/samsaradayplanner/ui/tasklistview/TaskListViewLogic.kt
package com.wiseassblog.samsaradayplanner.ui.tasklistview

import com.wiseassblog.samsaradayplanner.common.BaseViewLogic
import com.wiseassblog.samsaradayplanner.domain.ITaskStorage
import com.wiseassblog.samsaradayplanner.domain.constants.Messages.GENERIC_ERROR_MESSAGE

class TaskListViewLogic(
    private val view: ITaskListViewContract.View,
    private val vm: ITaskListViewContract.ViewModel,
    private val storage: ITaskStorage
) : BaseViewLogic<TasksListViewEvent>() {
    override fun onViewEvent(event: TasksListViewEvent) {
        when (event) {
            is TasksListViewEvent.OnStart -> onStart()
            is TasksListViewEvent.OnListItemSelected -> onItemSelected(event.position)
            is TasksListViewEvent.OnBackPressed -> onBackPressed()
        }
    }

    private fun onBackPressed() {
        view.navigateToDayView()
    }

    private fun onItemSelected(position: Int) {
        val id = vm.tasks.get()[position].taskId
        view.navigateToTaskView(id)
    }

    private fun onStart() {
        storage.getTasks(
            { tasks ->
                vm.tasks = tasks
                view.setTasks(tasks)
            },
            {
                view.showMessage(GENERIC_ERROR_MESSAGE)
                view.navigateToDayView()
            }
        )
    }
}